### PR TITLE
Dims and examples

### PR DESCRIPTION
**Pull request by wrobstory**
Learned a lot tonight around the state of the vega-lite API. In building for faceting, there is some very strange stuff around simple things like setting the chart width a height. They *really* need some APIs that abstract away things like setting categorical band widths and the `singleWidth`/`singleHeight` fields, and just allow a user to specify a chart width/height, and let vega figure out how to stage facets within it. 

I started a pretty straightforward tutorial notebook- working through things like building simple charts really finds the weird corners in the spec.

### CODE REVIEW

**Comment by jakevdp** [comment]
This is great – thanks for putting it together.

One minor comment: for the data, rather than list-of-dicts I'd prefer dict-of-arrays. While that's true to the internal storage format, I wouldn't want people reading this to think that they have to construct the data that way in Python!

**Comment by wrobstory** [comment]
Ok, done!

One thing we might want to consider is how much support we want for native Python data structures. Vincent supported lists of dicts, dicts-of-lists, numpy arrays, Series, and DataFrames, but honestly I have no idea how much they were ever used. Maybe we can say "everything needs to be a DataFrame".

**Comment by jakevdp** [comment]
I think it makes the most sense to support everything the new mpl named-data api does; basically either dataframes, dict of lists/arrays, or anything that "quacks like one of those", so to speak...

**Comment by tacaswell** [comment]
One of the big pluses of the duck-type is that we get support for lazy loading for almost free (ex h5py objects 'just work').

**Comment by ellisonbg** [comment]
Yeah, I am +1 on allowing any object that follows the matplotlib data kwarg
semantics. The only question is if we should always convert to a
pd.DataFrame once we get it. If we don't, then renderers may end up doing
it to manage the groupby's and aggregations. The only other option is
trying to to the groupbys with toolz.

On Thu, Sep 24, 2015 at 8:28 AM, Thomas A Caswell <notifications@github.com>
wrote:

> One of the big pluses of the duck-type is that we get support for lazy
> loading for almost free (ex h5py objects 'just work').
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ellisonbg/altair/pull/61#issuecomment-142963308>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by ellisonbg** [comment]
Ohh, one question - have you been able to get the width and height working?
Jeremy and I tried late the second night, but weren't able to. Seemed like
a bug with vega/vega-lite.

On Thu, Sep 24, 2015 at 8:40 AM, Brian Granger <ellisonbg@gmail.com> wrote:

> Yeah, I am +1 on allowing any object that follows the matplotlib data
> kwarg semantics. The only question is if we should always convert to a
> pd.DataFrame once we get it. If we don't, then renderers may end up doing
> it to manage the groupby's and aggregations. The only other option is
> trying to to the groupbys with toolz.
>
> On Thu, Sep 24, 2015 at 8:28 AM, Thomas A Caswell <
> notifications@github.com> wrote:
>
>> One of the big pluses of the duck-type is that we get support for lazy
>> loading for almost free (ex h5py objects 'just work').
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ellisonbg/altair/pull/61#issuecomment-142963308>.
>>
>
>
>
> --
> Brian E. Granger
> Associate Professor of Physics and Data Science
> Cal Poly State University, San Luis Obispo
> @ellisonbg on Twitter and GitHub
> bgranger@calpoly.edu and ellisonbg@gmail.com
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

File: altair/api.py [hunk 1]
```diff
@@ -282,6 +284,32 @@ def encode(self, **kwargs):
         self.encoding = Encoding(**kwargs)
         return self
 
+    def configure(self, **kwargs):
+        """Set chart configuration"""
+        self.vlconfig = VLConfig(**kwargs)
+        return self
+
+    def set_single_dims(self, width, height):
```
Comment (ellisonbg) [inline altair/api.py]: Maybe put this logic in a `_height_changed` and `_width_changed` method of the VLConfig class so the other things get updated automatically always. After playing with this, do you think a user will ever want to fine tune this stuff? Seems confusing enough that maybe not? We should definitely bring with up with Jeff.

**Comment by wrobstory** [comment]
Regarding width and height: I walked through this a little bit in the Notebook in this PR. Here are the width/height semantics for vega-lite: 

* `spec.config.width`/`spec.config.height` control the *figure* width/height, including all facets
* `spec.config.singleWidth`/`spec.config.singleHeight` control the individual *facet* width/height, except...
* If you have type `O` or `N` on an axis, the width/height of the figure are *entirely* determined by the `band.size` parameter for that specific encoding. For example, `spec.encoding.x.band.size` will always take precedent over `spec.config.singleWidth`. 

If the above is really frustrating for me (who has a tolerance for figure fiddling) to get charts rendering properly, it's going to be abysmal for the average user.

**Comment by ellisonbg** [comment]
Hmm, that sounds pretty wonky. What is the conceptual meaning of "band" in
this context?

On Thu, Sep 24, 2015 at 8:49 AM, Rob Story <notifications@github.com> wrote:

> Regarding width and height: I walked through this a little bit in the
> Notebook in this PR. Here are the width/height semantics for vega-lite:
>
>    - spec.config.width/spec.config.height control the *figure*
>    width/height, including all facets
>    - spec.config.singleWidth/spec.config.singleHeight control the
>    individual *facet* width/height, except...
>    - If you have type O or N on an axis, the width/height of the figure
>    are *entirely* determined by the band.size parameter for that specific
>    encoding. For example, spec.encoding.x.band.size will always take
>    precedent over spec.config.singleWidth.
>
> If the above is really frustrating for me (who has a tolerance for figure
> fiddling) to get charts rendering properly, it's going to be abysmal for
> the average user.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ellisonbg/altair/pull/61#issuecomment-142969696>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

File: altair/api.py (hunk 1 above)
Comment (wrobstory) [inline altair/api.py]: I worried a bit about making this behavior automatic for a couple reason:

* The relationship between width/height and singleWidth/singleHeight will then be largely implicit
* This only really works well about half the time (see all the manual fiddling in the Notebook in this PR). 

At this point, users have to understand the (complex) relationship between width/height, singleWidth/singleHeight, and `band.size` to get their charts rendering correctly. 

I think we *can* abstract this away, but its going to take a lot of work. For example, an aggregation can force the number of ordinal/nominal bands that get created, and then the user will need to fiddle with `band.size` to get the exact dimensioning they need.

**Comment by wrobstory** [comment]
> What is the conceptual meaning of "band" in
this context?

A band is a single ordinal or nominal unit. You're basically saying "make each O/N unit x pixels wide"

**Comment by ellisonbg** [comment]
Ahh, I understand now and can see why this will be horribly confusing for
users. Also, I think this way of sizing makes little sense for renderers
other than the vega one. We should work with Jeff to make sure that the
implementation details of vega/vega-lite are not leaking out into the spec.

On Thu, Sep 24, 2015 at 8:55 AM, Rob Story <notifications@github.com> wrote:

> What is the conceptual meaning of "band" in
> this context?
>
> A band is a single ordinal or nominal unit. You're basically saying "make
> each O/N unit x pixels wide"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ellisonbg/altair/pull/61#issuecomment-142971769>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by ellisonbg** [comment]
Let's merge this and iterate on the width/height ideas then...

On Thu, Sep 24, 2015 at 9:35 AM, Brian Granger <ellisonbg@gmail.com> wrote:

> Ahh, I understand now and can see why this will be horribly confusing for
> users. Also, I think this way of sizing makes little sense for renderers
> other than the vega one. We should work with Jeff to make sure that the
> implementation details of vega/vega-lite are not leaking out into the spec.
>
> On Thu, Sep 24, 2015 at 8:55 AM, Rob Story <notifications@github.com>
> wrote:
>
>> What is the conceptual meaning of "band" in
>> this context?
>>
>> A band is a single ordinal or nominal unit. You're basically saying "make
>> each O/N unit x pixels wide"
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ellisonbg/altair/pull/61#issuecomment-142971769>.
>>
>
>
>
> --
> Brian E. Granger
> Associate Professor of Physics and Data Science
> Cal Poly State University, San Luis Obispo
> @ellisonbg on Twitter and GitHub
> bgranger@calpoly.edu and ellisonbg@gmail.com
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Pull request merged**